feat: use output from new doc-ci with odoc 3 features in the package documentation area

panglesd opened the pull request:
Replaces #3123, but PR is on `main`.

This PR uses the output from the new doc-ci @jonludlam has been working on. The docs are built with odoc 3, benefiting from several features, some of them that were previously "emulated" on this repo (such as a global sidebar).

This currently implements:
- [x] A global sidebar as computed by odoc 3
- [x] Better breadcrumbs
- [x] Source code pages
- [x] Redirections for url changes

This PR is not yet complete, however all pushed commits can already be reviewed, so I think it was already worth opening. 
I tried to make the commits as atomic as possible to ease the review, which can then be made "commit by commit". All commits compile, although they may fail at runtime (eg due to wrong format for deserializing).

I'll post updates and warn when I consider the PR complete!

panglesd committed:
Get docs from new docs CI

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: Update to new odoc3 json format

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: use new `status.json` from odoc 3

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: use header field

It was added in https://github.com/ocaml/odoc/pull/1314 and needs not be ignored

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: fix search script location change with odoc_driver

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: fix search results' url

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: compute sidebar from toc json

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: do not reload global toc on toc click

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: uniformize toc look

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: small refactor of breadcrumbs

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs: allow breadcrumbs without href

Odoc 3 allows to have breadcrumbs without href. Currently, the breadcrumbs are
still computed by ocaml.org, but soon it will be taken from odoc3's json output.

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs breadcrumbs: Adding `page` kind

With odoc 3 and hierarchical documentation, breadcrumbs will include pages.

They are rendered with `/` between them (while other components are rendered
with `.` between them). So We can have `library1/Module.Submodule`.

Pages cannot yet happen since we still compute breadcrumbs ourselves.

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
rename `library_path` to `path`: pages also have/are in breadcrumbs

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Docs breadcrumbs: use the one computed by odoc without processing

Co-authored-by: Jon Ludlam <jon@recoil.org>

panglesd committed:
Add dependency on ppx_deriving_yojson

panglesd committed:
Docs: hide toc until alpine has loaded

From https://alpinejs.dev/directives/cloak :

> Sometimes, when you're using AlpineJS for a part of your template, there is a
> "blip" where you might see your uninitialized template after the page loads,
> but before Alpine loads.
>
> `x-cloak` addresses this scenario by hiding the element it's attached to until
> Alpine is fully loaded on the page.
>
> For `x-cloak` to work however, you must add the following CSS to the page.
>
> ```
> [x-cloak] { display: none !important; }
> ```

sabine commented:
I converted this to a draft, but feel free to ping me any time when it makes sense to test this. I will put this on https://staging.ocaml.org, and hook it up with the staging docs-ci.

panglesd committed:
Add source pages to docs

yawaramin commented:
Does this make #1963 obsolete?

jonludlam commented:
> Does this make #1963 obsolete?

Yes.

panglesd committed:
Render html in breadcrumbs

panglesd committed:
Add style for rendered source code

panglesd committed:
Handle redirection

Since odoc 3, and the replacement of voodoo by odoc_driver as the odoc driver,
the layout for docs has changed. Since we do not want to break previously
existing links, we redirect the old layout to the new one.

panglesd committed:
Sidebar: open sub-toc when clicking on a page from the sidebar

panglesd commented:
@sabine @jonludlam This is ready for testing on staging!

(I don't know what makes the "deployability" CI fail but I don't think it's related)

sabine commented:
Deploying on staging.ocaml.org (https://deploy.ci.ocaml.org/job/2025-06-06/074257-ocluster-build-384752)

jonludlam commented:
The source rendering needs some CSS love:

<img width="1065" alt="Screenshot 2025-06-09 at 14 43 27" src="https://github.com/user-attachments/assets/885ae0e3-6034-4e69-8af3-e4bdace5be28" />

panglesd commented:
Did you take the screenshot from staging.ocaml.org? Here is how the [same page](https://staging.ocaml.org/p/base/v0.17.2/doc/src/base/char.ml.html) looks for me:

Light mode:
![image](https://github.com/user-attachments/assets/6b260783-8f06-447e-819f-6d93251927bd)

Dark mode:
![image](https://github.com/user-attachments/assets/1c197a9b-4e03-4e8d-a2d1-56d5250b510b)


I'll try to reproduce with different browser and settings, but if you have any hint, I'll take them.

panglesd committed:
Doc: handle asset loading

Documentation assets are not present as json files. Therefore, if we do not find
a .json target, we try to load it as an asset.

This is not ideal: assets always require two requests to the documentation
server. 404 pages also trigger two requests...

panglesd committed:
Fix package files' links

panglesd committed:
Fix package files appearing in global toc

panglesd commented:
Ok, testing on chromium yields the same CSS problem...

sabine commented:
Deploying updated PR to staging.

panglesd committed:
Fix theme for source rendering

panglesd committed:
Revert keeping the sidebar on new page loading

panglesd committed:
Finally implement avoid reloading sidebar

panglesd committed:
Docs: Improve source code target highlighting

panglesd committed:
Improve css of source rendering line numbers

panglesd committed:
Improve links to source style

panglesd committed:
Docs: only show right sidebar when it exists

panglesd committed:
Improve error handling for status.json

sabine commented on src/ocamlorg_web/lib/handler.ml:
```diff
@@ -1269,95 +1281,82 @@ let package_documentation t kind req =
+              | Some "class-type" -> Class_type
+              | Some "file" -> File
+              | Some "source" -> Source
+              | None -> Page
```
makes sense to do this decoding in the `Sidebar` module in `ocamlorg_package/`, instead of here

panglesd committed:
Fix enter on search result

Odoc 3's search has already a leading `/`.

panglesd committed:
Fix redirection on edge cases

`old_path` and `new_path` do not include the prefix, which makes the detection
of redirection harder.

Previously, the url `/p/odoc/3.0.0/doc/odoc.odoc/Odoc_odoc/Sidebar/index.html`
would trigger the rewriting:
`doc/Odoc_odoc/Sidebar/index.html`
->
`doc/odoc.odoc/Odoc_odoc/Sidebar/index.html`

which is a bug. Adding a / in the rewriting rules solves it

panglesd committed:
Put documentation-related code in a documentation file

panglesd committed:
Consider non-html files as assets

panglesd committed:
Turn sidebar into Navmap earlier

panglesd committed:
Increment cache version and update comments around it

shonfeder commented:
Hello! Congrats on this work, @panglesd!

@sabine, just FYI, I will be taking over responsibility for helping ensure this lands.

If any additional work is needed here, I'll respond, and if not I may ping with followups until it is in :)

Thanks!

jonludlam commented:
Can we please not merge this PR for a few days - I just need to fiddle with the storage on the new server and I'd rather not do that when it's being used! I'll update when the changes have been done.

jonludlam commented:
OK, sage has been updated, we're good to go!

cuihtlauac commented:
Excellent @jonludlam. Do you think we can wait until Monday to press the merge button? I hate Friday releases :-)

jonludlam commented:
Certainly, and I agree, releasing on a Friday afternoon is not the best time :-)

shonfeder commented:
Hello! Are we good to merge here, now that it is Monday? Thanks :)

cuihtlauac commented:
Sorry, I missed this week's Monday.
